xarray/namedarray/core.py

```diff
@@ -819,11 +819,18 @@ def chunk(
                 if dim in chunks
             }
 
-        chunkmanager = guess_chunkmanager(chunked_array_type)
-
         data_old = self._data
-        if chunkmanager.is_chunked_array(data_old):
-            data_chunked = chunkmanager.rechunk(data_old, chunks)  # type: ignore[arg-type]
+        if is_chunked_array(data_old):
+            print(f"problematic chunks = {chunks}")
+            # if is_dict_like(chunks) and chunks != {}:
+            #     chunks = tuple(chunks.get(n, s) for n, s in enumerate(data_old.shape))  # type: ignore[assignment]
+
+            print(f"hopefully normalized chunks = {chunks}")
```

Comment:
This is really irritating - if I keep these lines commented out then my `test_rechunk` on the `DummyChunkedArray` fails. But if I uncomment these lines (therefore doing exactly what happens in the other branch of the `if is_chunked_array(data_old):` statement) then dask rechunk tests fail! 

There are so many possible valid argument types for `chunks` here, some of which are `dicts` but completely different, e.g. `{0: (2, 3)}` vs `{'x': (2, 3)}`.

It would be much nicer for all possible `chunks` to go through a single `normalize_chunks` function, but I'm getting confused even trying to work out what the current behaviour is.

The `ChunkManager` has a `.normalize_chunks` method, to call out to `dask.array.normalize_chunks`. Cubed [vendors this function](https://github.com/cubed-dev/cubed/blob/69e9f94241f3b64924f52ec50e0548c24f79a8b6/cubed/vendor/dask/array/core.py#L103) too, so perhaps instead xarray should vendor `dask.array.normalize_chunks` and remove it from the `ChunkManager` class?

Reply:
> `test_rechunk` on the `DummyChunkedArray` fails

This was actually mostly my fault for having a bug in that test, fixed by https://github.com/pydata/xarray/pull/9286/commits/0296f92d207afb7319b4e31ff03a76c145ddb251.

> It would be much nicer for all possible chunks to go through a single normalize_chunks function

But there is still some unnecessary complexity that would be nice to remove. The main reason why the weird `is_dict_like(chunks):` sections that turn dicts of chunks into tuples are currently needed is because of this bug in `dask.array.core.normalize_chunks` https://github.com/dask/dask/issues/11261. Otherwise we could just use that.

(If we do just use that we should perhaps vendor it though - as cubed does already).

Reply:
I managed to sort this all out, so now everything goes through `dask.array.core.normalize_chunks`, which is much neater.

Question is now do I:
1) Vendor `dask.array.core.normalize_chunks` (like cubed does), and use the vendored version no matter which `ChunkManager` is called
2) Make all chunkmanagers define a normalize_chunks method and refer to that (what the `main` code currently does).

I think we actually have to do (1), because we now have a codepath which will try to call `normalize_chunks` even on chunked arrays that do not define a chunkmanager. But we want to vendor it without introducing any more dependencies (e.g. `toolz`).

Reply:
@dcherian I would appreciate your input on this vendoring question before I move ahead with it ^

Reply:
Vendor it! Sorry for the delay. We can generalize if it's ever needed